python/pyspark/mllib/linalg/__init__.py

```diff
@@ -1296,9 +1296,19 @@ def asML(self):
         return newlinalg.SparseMatrix(self.numRows, self.numCols, self.colPtrs, self.rowIndices,
                                       self.values, self.isTransposed)
 
-    # TODO: More efficient implementation:
     def __eq__(self, other):
-        return np.all(self.toArray() == other.toArray())
+        if (not isinstance(other, SparseMatrix) or
```

Comment:
You may have some issues with subclasses if you do it this way (the equality operation is not commutative) One recommended way is `if type(other) is type(self):...`